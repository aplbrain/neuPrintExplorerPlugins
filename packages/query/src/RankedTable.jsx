/*
 * Implements table view that shows ordered strongest connection to each neuron
 * and visually indicates the different classes of neurons.  (This is meant
 * to be similar to Lou's tables.)
 */
import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';
import randomColor from 'randomcolor';

import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import { withStyles } from '@material-ui/core/styles';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import FormLabel from '@material-ui/core/FormLabel';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Typography from '@material-ui/core/Typography';

import ColorBox from './visualization/ColorBox';
import NeuronHelp from './shared/NeuronHelp';

const styles = () => ({
  textField: {},
  formControl: {
    margin: '0.5em 0 1em 0',
    width: '100%'
  }
});

// available colors

/* const colorArray = [
  '#8dd3c7',
  '#ffffb3',
  '#bebada',
  '#fb8072',
  '#80b1d3',
  '#fdb462',
  '#b3de69',
  '#fccde5',
  '#d9d9d9',
  '#bc80bd',
  '#ccebc5',
  '#ffed6f'
]; */

const pluginName = 'RankedTable';
const pluginAbbrev = 'rt';

class RankedTable extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      neuronSrc: '',
      preOrPost: 'pre'
    };
  }

  static get queryName() {
    return 'Ranked Table';
  }

  static get queryDescription() {
    return 'Show connections to neuron(s) ranked in order and colored by neuron class';
  }

  static get queryAbbreviation() {
    return pluginAbbrev;
  }

  static get isExperimental() {
    return true;
  }

  processResults = (query, apiResponse) => {
    const { dataSet } = query;

    const colorMap = {};
    const reverseCounts = {};

    const data = [];
    let lastBody = -1;
    let columns = [];
    let maxColumns = 0;

    apiResponse.data.forEach(row => {
      const [, , weight, body2, , , mId, nId, preId, body1] = row;

      if (
        (query.parameters.find_inputs === false && (preId !== mId || nId === mId)) ||
        (query.parameters.find_inputs === true && preId === mId)
      ) {
        if (body2 in reverseCounts) {
          reverseCounts[String(body2)][String(body1)] = weight;
        } else {
          reverseCounts[String(body2)] = {};
          reverseCounts[String(body2)][String(body1)] = weight;
        }
      }
    });

    apiResponse.data.forEach((row, index) => {
      const [neuron1, neuron2, weight, body2, , neuron2Type, mId, nId, preId, body1] = row;

      if (
        (query.parameters.find_inputs === false && preId === mId) ||
        (query.parameters.find_inputs === true && (preId !== mId || nId === mId))
      ) {
        // check the colormap for the current type.
        // if present use the existing color.
        // else add a new one from the list. Unless they have all been used,
        // then add a random one.
        if (!(neuron2Type in colorMap)) {
          colorMap[neuron2Type] = randomColor({ luminosity: 'light', hue: 'random' });
        }
        // color should be white if we don't have a type defined.
        const weightColor = neuron2Type ? colorMap[neuron2Type] : '#ffffff';

        // start a new row for each body1.
        if (body1 !== lastBody) {
          data.push(columns);
          maxColumns = Math.max(columns.length, maxColumns);
          columns = [`${neuron1 || body1} (${body1})`];
        }

        let reverseWeight = 0;
        if (body2 in reverseCounts && body1 in reverseCounts[body2]) {
          reverseWeight = reverseCounts[body2][body1];
        }

        const cellKey = `${index}${body1}${body2}`;

        // add the current item to the current columns array
        columns.push({
          value: (
            <ColorBox
              margin={0}
              width={85}
              height={85}
              backgroundColor={weightColor}
              title=""
              key={cellKey}
              text={
                <div>
                  <Typography>{neuron2 || body2}</Typography>
                  <Typography variant="caption">{weight}</Typography>
                  <Typography variant="caption">{reverseWeight}</Typography>
                </div>
              }
            />
          ),
          uniqueId: cellKey
        });

        lastBody = body1;
      }
    });
    // make sure the last column is added.
    maxColumns = Math.max(columns.length, maxColumns);
    data.push(columns);

    const headings = Array(maxColumns - 1)
      .fill(0)
      .map((e, i) => `#${i + 1}`);

    return {
      columns: ['', ...headings],
      data,
      debug: apiResponse.debug,
      dataSet
    };
  };

  processRequest = () => {
    const { neuronSrc, preOrPost } = this.state;
    const { dataSet, actions, history } = this.props;
    if (neuronSrc !== '') {
      const parameters = { dataset: dataSet };
      if (/^\d+$/.test(neuronSrc)) {
        parameters.neuron_id = parseInt(neuronSrc, 10);
      } else {
        parameters.neuron_name = neuronSrc;
      }
      let title = `Outputs from ${neuronSrc}`;

      if (preOrPost === 'pre') {
        parameters.find_inputs = false;
      } else {
        parameters.find_inputs = true;
        title = `Inputs to ${neuronSrc}`;
      }

      const query = {
        dataSet,
        queryString: '/npexplorer/rankedtable',
        visType: 'HeatMapTable',
        visProps: { squareSize: 85 },
        plugin: pluginName,
        parameters,
        title,
        menuColor: randomColor({ luminosity: 'light', hue: 'random' }),
        processResults: this.processResults
      };

      actions.submit(query);
      history.push({
        pathname: '/results',
        search: actions.getQueryString()
      });
    } else {
      actions.formError('Please enter a neuron name.');
    }
  };

  addNeuron = event => {
    this.setState({ neuronSrc: event.target.value });
  };

  setDirection = event => {
    this.setState({ preOrPost: event.target.value });
  };

  catchReturn = event => {
    // submit request if user presses enter
    if (event.keyCode === 13) {
      event.preventDefault();
      this.processRequest();
    }
  };

  render() {
    const { classes, isQuerying } = this.props;
    const { neuronSrc, preOrPost } = this.state;
    return (
      <div>
        <FormControl className={classes.formControl}>
          <NeuronHelp>
            <TextField
              label="Neuron name"
              multiline
              fullWidth
              rows={1}
              value={neuronSrc}
              rowsMax={4}
              className={classes.textField}
              onChange={this.addNeuron}
              onKeyDown={this.catchReturn}
            />
          </NeuronHelp>
        </FormControl>
        <FormControl component="fieldset" required className={classes.formControl}>
          <FormLabel component="legend">Neuron Direction</FormLabel>
          <RadioGroup
            aria-label="preorpost"
            name="preorpost"
            className={classes.group}
            value={preOrPost}
            onChange={this.setDirection}
          >
            <FormControlLabel
              value="pre"
              control={<Radio color="primary" />}
              label="Pre-synaptic"
            />
            <FormControlLabel
              value="post"
              control={<Radio color="primary" />}
              label="Post-synaptic"
            />
          </RadioGroup>
        </FormControl>
        <Button
          disabled={isQuerying}
          variant="contained"
          color="primary"
          onClick={this.processRequest}
        >
          Submit
        </Button>
      </div>
    );
  }
}

RankedTable.propTypes = {
  isQuerying: PropTypes.bool.isRequired,
  history: PropTypes.object.isRequired,
  actions: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired,
  dataSet: PropTypes.string.isRequired
};

export default withRouter(withStyles(styles)(RankedTable));
