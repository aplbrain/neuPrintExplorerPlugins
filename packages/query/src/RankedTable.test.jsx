import { RankedTable } from './RankedTable';

const styles = { select: {}, clickable: {} };
const { actions, React, renderer } = global;

const neoServerSettings = {
  get: () => 'http://example.com'
};

const component = (
  <RankedTable
    availableROIs={['roiA', 'roiB', 'roiC']}
    dataSet="test"
    datasetstr="test"
    actions={actions}
    submit={actions.submit}
    classes={styles}
    history={{ push: jest.fn() }}
    isQuerying={false}
    neoServerSettings={neoServerSettings}
    neoServer="testServer"
    isPublic={false}
  />
);

const componentPublic = (
  <RankedTable
    availableROIs={['roiA', 'roiB', 'roiC']}
    dataSet="test"
    datasetstr="test"
    actions={actions}
    submit={actions.submit}
    classes={styles}
    history={{ push: jest.fn() }}
    isQuerying={false}
    neoServerSettings={neoServerSettings}
    neoServer="testServer"
    isPublic
  />
);

describe('RankedTable Plugin', () => {
  it('should have required details fields', () => {
    expect(RankedTable.details.name).toBeTruthy();
    expect(RankedTable.details.description).toBeTruthy();
  });
  it('renders correctly', () => {
    const pluginView = renderer.create(component).toJSON();
    expect(pluginView).toMatchSnapshot();
  });
  it('renders correctly in public mode', () => {
    const pluginViewPublic = renderer.create(componentPublic).toJSON();
    expect(pluginViewPublic).toMatchSnapshot();
  });
});
