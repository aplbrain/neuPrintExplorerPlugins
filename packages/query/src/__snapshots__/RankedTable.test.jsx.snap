// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RankedTable Plugin renders correctly 1`] = `
<div>
  <div
    className="MuiFormControl-root-1"
  >
    <div>
      <div
        className="MuiFormControl-root-1 MuiFormControl-fullWidth-4"
        onKeyDown={[Function]}
      >
        <label
          className="MuiFormLabel-root-17 MuiInputLabel-root-6 MuiInputLabel-formControl-11 MuiInputLabel-animated-14"
          data-shrink={false}
        >
          Neuron name
        </label>
        <div
          className="MuiInputBase-root-37 MuiInput-root-24 MuiInput-underline-28 MuiInputBase-fullWidth-46 MuiInput-fullWidth-31 MuiInputBase-formControl-38 MuiInput-formControl-25 MuiInputBase-multiline-45 MuiInput-multiline-30"
          onClick={[Function]}
        >
          <div
            className="MuiPrivateTextarea-root-54"
          >
            <textarea
              aria-hidden="true"
              className="MuiPrivateTextarea-textarea-55 MuiPrivateTextarea-shadow-56"
              readOnly={true}
              rows="1"
              tabIndex={-1}
              value=""
            />
            <textarea
              aria-hidden="true"
              className="MuiPrivateTextarea-textarea-55 MuiPrivateTextarea-shadow-56"
              readOnly={true}
              rows={1}
              tabIndex={-1}
              value=""
            />
            <textarea
              aria-invalid={false}
              className="MuiPrivateTextarea-textarea-55 MuiInputBase-input-47 MuiInput-input-32 MuiInputBase-inputMultiline-49 MuiInput-inputMultiline-34"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              rows={1}
              style={
                Object {
                  "height": 19,
                }
              }
              value=""
            />
          </div>
        </div>
      </div>
      <p
        aria-describedby={null}
        className="MuiTypography-root-65 MuiTypography-body1-74 MuiTypography-colorError-99 NeuronHelp-badge-5"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        title="Enter body ID, neuron name, or wildcard names using period+star (e.g., MBON.*). Warning: if using regular expressions, special characters like parentheses must be escaped (e.g. Delta6g\\\\\\\\(preQ7\\\\\\\\).*)"
      >
        ?
      </p>
    </div>
  </div>
  <fieldset
    className="MuiFormControl-root-1"
  >
    <div
      aria-label="preorpost"
      className="MuiFormGroup-root-100"
      role="radiogroup"
    >
      <label
        className="MuiFormControlLabel-root-102"
      >
        <span
          className="MuiButtonBase-root-123 MuiIconButton-root-117 MuiPrivateSwitchBase-root-113 MuiRadio-root-108 MuiRadio-colorPrimary-111 MuiPrivateSwitchBase-checked-114 MuiRadio-checked-109"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label-122"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-126"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <input
              checked={true}
              className="MuiPrivateSwitchBase-input-116"
              disabled={false}
              name="preorpost"
              onChange={[Function]}
              type="radio"
              value="pre"
            />
          </span>
          <span
            className="MuiTouchRipple-root-170"
          />
        </span>
        <span
          className="MuiTypography-root-65 MuiTypography-body1-74 MuiFormControlLabel-label-107"
        >
          Rank outputs
        </span>
      </label>
      <label
        className="MuiFormControlLabel-root-102"
      >
        <span
          className="MuiButtonBase-root-123 MuiIconButton-root-117 MuiPrivateSwitchBase-root-113 MuiRadio-root-108 MuiRadio-colorPrimary-111"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label-122"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-126"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <input
              checked={false}
              className="MuiPrivateSwitchBase-input-116"
              disabled={false}
              name="preorpost"
              onChange={[Function]}
              type="radio"
              value="post"
            />
          </span>
          <span
            className="MuiTouchRipple-root-170"
          />
        </span>
        <span
          className="MuiTypography-root-65 MuiTypography-body1-74 MuiFormControlLabel-label-107"
        >
          Rank inputs
        </span>
      </label>
    </div>
  </fieldset>
  <div
    className="MuiFormControl-root-1"
  >
    <label
      className="MuiFormControlLabel-root-102"
    >
      <span
        className="MuiSwitch-root-135"
      >
        <span
          className="MuiButtonBase-root-123 MuiIconButton-root-117 MuiPrivateSwitchBase-root-113 MuiSwitch-switchBase-138 MuiSwitch-colorPrimary-140"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label-122"
          >
            <span
              className="MuiSwitch-icon-136"
            />
            <input
              checked={false}
              className="MuiPrivateSwitchBase-input-116"
              disabled={false}
              onChange={[Function]}
              type="checkbox"
            />
          </span>
          <span
            className="MuiTouchRipple-root-170"
          />
        </span>
        <span
          className="MuiSwitch-bar-143"
        />
      </span>
      <span
        className="MuiTypography-root-65 MuiTypography-body1-74 MuiFormControlLabel-label-107"
      >
        <p
          className="MuiTypography-root-65 MuiTypography-body1-74"
          style={
            Object {
              "display": "inline-flex",
            }
          }
        >
          Limit to high-confidence synapses
        </p>
      </span>
    </label>
  </div>
  <button
    className="MuiButtonBase-root-123 MuiButton-root-144 MuiButton-contained-155 MuiButton-containedPrimary-156 MuiButton-raised-158 MuiButton-raisedPrimary-159"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-145"
    >
      Submit
    </span>
    <span
      className="MuiTouchRipple-root-170"
    />
  </button>
</div>
`;

exports[`RankedTable Plugin renders correctly in public mode 1`] = `
<div>
  <div
    className="MuiFormControl-root-1"
  >
    <div>
      <div
        className="MuiFormControl-root-1 MuiFormControl-fullWidth-4"
        onKeyDown={[Function]}
      >
        <label
          className="MuiFormLabel-root-17 MuiInputLabel-root-6 MuiInputLabel-formControl-11 MuiInputLabel-animated-14"
          data-shrink={false}
        >
          Neuron name
        </label>
        <div
          className="MuiInputBase-root-37 MuiInput-root-24 MuiInput-underline-28 MuiInputBase-fullWidth-46 MuiInput-fullWidth-31 MuiInputBase-formControl-38 MuiInput-formControl-25 MuiInputBase-multiline-45 MuiInput-multiline-30"
          onClick={[Function]}
        >
          <div
            className="MuiPrivateTextarea-root-54"
          >
            <textarea
              aria-hidden="true"
              className="MuiPrivateTextarea-textarea-55 MuiPrivateTextarea-shadow-56"
              readOnly={true}
              rows="1"
              tabIndex={-1}
              value=""
            />
            <textarea
              aria-hidden="true"
              className="MuiPrivateTextarea-textarea-55 MuiPrivateTextarea-shadow-56"
              readOnly={true}
              rows={1}
              tabIndex={-1}
              value=""
            />
            <textarea
              aria-invalid={false}
              className="MuiPrivateTextarea-textarea-55 MuiInputBase-input-47 MuiInput-input-32 MuiInputBase-inputMultiline-49 MuiInput-inputMultiline-34"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              rows={1}
              style={
                Object {
                  "height": 19,
                }
              }
              value=""
            />
          </div>
        </div>
      </div>
      <p
        aria-describedby={null}
        className="MuiTypography-root-65 MuiTypography-body1-74 MuiTypography-colorError-99 NeuronHelp-badge-5"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        title="Enter body ID, neuron name, or wildcard names using period+star (e.g., MBON.*). Warning: if using regular expressions, special characters like parentheses must be escaped (e.g. Delta6g\\\\\\\\(preQ7\\\\\\\\).*)"
      >
        ?
      </p>
    </div>
  </div>
  <fieldset
    className="MuiFormControl-root-1"
  >
    <div
      aria-label="preorpost"
      className="MuiFormGroup-root-100"
      role="radiogroup"
    >
      <label
        className="MuiFormControlLabel-root-102"
      >
        <span
          className="MuiButtonBase-root-123 MuiIconButton-root-117 MuiPrivateSwitchBase-root-113 MuiRadio-root-108 MuiRadio-colorPrimary-111 MuiPrivateSwitchBase-checked-114 MuiRadio-checked-109"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label-122"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-126"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <input
              checked={true}
              className="MuiPrivateSwitchBase-input-116"
              disabled={false}
              name="preorpost"
              onChange={[Function]}
              type="radio"
              value="pre"
            />
          </span>
          <span
            className="MuiTouchRipple-root-170"
          />
        </span>
        <span
          className="MuiTypography-root-65 MuiTypography-body1-74 MuiFormControlLabel-label-107"
        >
          Rank outputs
        </span>
      </label>
      <label
        className="MuiFormControlLabel-root-102"
      >
        <span
          className="MuiButtonBase-root-123 MuiIconButton-root-117 MuiPrivateSwitchBase-root-113 MuiRadio-root-108 MuiRadio-colorPrimary-111"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label-122"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-126"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <input
              checked={false}
              className="MuiPrivateSwitchBase-input-116"
              disabled={false}
              name="preorpost"
              onChange={[Function]}
              type="radio"
              value="post"
            />
          </span>
          <span
            className="MuiTouchRipple-root-170"
          />
        </span>
        <span
          className="MuiTypography-root-65 MuiTypography-body1-74 MuiFormControlLabel-label-107"
        >
          Rank inputs
        </span>
      </label>
    </div>
  </fieldset>
  
  <button
    className="MuiButtonBase-root-123 MuiButton-root-144 MuiButton-contained-155 MuiButton-containedPrimary-156 MuiButton-raised-158 MuiButton-raisedPrimary-159"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-145"
    >
      Submit
    </span>
    <span
      className="MuiTouchRipple-root-170"
    />
  </button>
</div>
`;
